modules/apps/frontend-js/frontend-js-components-web/src/main/resources/META-INF/resources/walkthrough/Walkthrough.js

```diff
@@ -0,0 +1,422 @@
+ * We should remove guesses from the top since people will not place things on top that will collide at the beginning of the viewport.
+ */
+const ALIGNMENTS_GUESS_MAP = {
+	...ALIGNMENTS_INVERSE_MAP,
+	blbr: 'top',
+	brbl: 'top',
+	clcr: 'left',
+	crcl: 'right',
+	tcbc: 'top',
+	tlbl: 'top',
+	trbr: 'top',
+};
+
+const WalkthroughOverlay = ({popoverVisible, trigger}) => {
+	const [overlayBounds, setOverlayBounds] = useState({
+		height: 0,
+		width: 0,
+		x: 0,
+		y: 0,
+	});
+
+	const updateOverlayBounds = useCallback(
+		(overlayBounds) => {
+			overlayBounds = overlayBounds
+				? overlayBounds
+				: trigger.getBoundingClientRect();
+
+			setOverlayBounds({
+				height: overlayBounds.height,
+				width: overlayBounds.width,
+				x: overlayBounds.x,
+				y: overlayBounds.y,
+			});
+		},
+		[trigger]
+	);
+
+	useEffect(() => {
+		updateOverlayBounds();
+	}, [updateOverlayBounds]);
+
+	useObserveRect(updateOverlayBounds, trigger);
+
+	return (
+		<OverlayMask
+			bounds={overlayBounds}
+			onBoundsChange={setOverlayBounds}
+			visible={popoverVisible}
+		/>
+	);
+};
+
+const WalkthroughStep = ({
+	closeOnClickOutside,
+	closeable,
+	content,
+	darkbg,
+	onNext,
+	onPrevious,
+	positioning: defaultPositioning = 'right-top',
+	previousTrigger,
+	skippable,
+	stepIndex,
+	stepLength,
+	title,
+	trigger,
+}) => {
+	const popoverRef = useRef(null);
+
+	const hotspotRef = useRef(null);
+
+	const [popoverVisible, setPopoverVisible] = useState(false);
+
+	const [currentAlignment, setCurrentAlignment] = useState(
+		defaultPositioning
+	);
+
+	/**
+	 * This useEffect was necessary because Walkthrough(Step) components
+	 * are always mounted. So, currentAligment will not be updated when
+	 * the component will be updated.
+	 */
+	useEffect(() => {
+		setCurrentAlignment(defaultPositioning);
+	}, [defaultPositioning]);
+
+	const align = useCallback(() => {
+		if (popoverVisible && popoverRef.current) {
+			const points = ALIGNMENTS_MAP[currentAlignment];
+
+			const alignment = doAlign({
+				offset: [OVERLAY_OFFSET_X, OVERLAY_OFFSET_Y],
+				overflow: {
+					adjustX: true,
+					adjustY: true,
+				},
+				points,
+				sourceElement: popoverRef.current,
+				targetElement: trigger,
+			});
+
+			const alignmentString = alignment.points.join('');
+
+			const pointsString = points.join('');
+
+			if (alignment.overflow.adjustX) {
+				setCurrentAlignment(ALIGNMENTS_GUESS_MAP[alignmentString]);
+			}
+			else if (pointsString !== alignmentString) {
+				setCurrentAlignment(ALIGNMENTS_INVERSE_MAP[alignmentString]);
+			}
+
+			if (!darkbg && previousTrigger !== trigger) {
+				trigger.classList.add('lfr-walkthrough-element-shadow');
+
+				previousTrigger.classList.remove(
+					'lfr-walkthrough-element-shadow'
+				);
+			}
+		}
+	}, [
+		currentAlignment,
+		darkbg,
+		popoverRef,
+		popoverVisible,
+		previousTrigger,
+		trigger,
+	]);
+
+	useEffect(() => {
+		align();
+	}, [align]);
```

Comment:
Same pattern of https://github.com/liferay-frontend/liferay-portal/pull/2164/files#r867263769